---
title: "LPS Bulk RNAseq DEG Volcano"
author: "Kennedi Todd"
date: "1/12/2022"
output: html_document
---

# Set working directory
```{r}
knitr::opts_knit$set(
  root.dir = "~/Dropbox/Mac/Documents/Shiny/finalized_apps/LPS_bulk_volcano")
```

# Load packages
```{r}
library(shiny)
library(stringr)
library(ggplot2)
library(ggrepel)
```

# Notes
* It's faster to load a rds file of a table than reading in the read.table()
```{r}
ptm <- proc.time()
data <- readRDS("LPS_brain_gene_table.rds")
proc.time() - ptm
```

# Test volcano
## Generate cutoff combinations
```{r}
FDRq_values <- c("0.01", "0.05", "0.10")
LFC_values <- c("0.0", "0.5", "1.0")
combinations <- vector()

for(i in FDRq_values){
  for(j in LFC_values) {
    string <- paste0("FDRq_", i, "_", "LFC_", j)
    combinations <- c(combinations, string)
  }
}

combinations
```

## Color based on cutoff
Adding color columns to table based on varying cutoffs.
```{r}
# loop through combinations
for (cutoff in combinations) {
  
  # initialize variables
  color_values <- vector()
  max <- nrow(data)
  num_DEGs <- vector()
  
  # parse cutoff info
  cutoff_values <- str_match(cutoff, "FDRq_(.+)_LFC_(.+)")
  FDRq <- as.numeric(cutoff_values[,2])
  LFC <- as.numeric(cutoff_values[,3])
  
  # loop through every gene
  for(i in 1:max){
    # if FDRq is met
    if (data$adj.P.Val[i] < FDRq){
      
      if (data$logFC[i] > LFC){
        color_values <- c(color_values, 1) # 1 when logFC met and positive and FDRq met
        num_DEGs <- c(num_DEGs,"up")
      }
      else if (data$logFC[i] < -LFC){
        color_values <- c(color_values, 2) # 2 when logFC met and negative and FDRq met
        num_DEGs <- c(num_DEGs, "down")
      }
      else{
        color_values <- c(color_values, 3) # 3 when logFC not met and FDRq met
        num_DEGs <- c(num_DEGs, "neither")
      }
    }
    # if FDRq is not met
    else{
      color_values <- c(color_values, 3) # 3 when FDRq not met
      num_DEGs <- c(num_DEGs, "neither")
    }
  }
  
  # add column to data
  data[cutoff] <- color_values
}
```

## User defined variables
```{r}
tissue <- "Brain"
cutoff <- "FDRq_0.05_LFC_1.0"
```

## Plot
```{r assign_colors}
# color column is equal to cutoff column
data$color <- factor(data[,cutoff])

# get FDRq and LFC values
FDRq <- as.numeric(str_match(cutoff,"FDRq_(.+)_LFC_(.+)")[,2])
LFC <- as.numeric(str_match(cutoff,"FDRq_(.+)_LFC_(.+)")[,3])

# Subset top 10 up and down regulated genes
up <- data[data$color == 1,]
up10 <- up[1:10,]
down <- data[data$color == 2,]
down10 <- down[1:10,]

# goi
goi <- c("IRF1,BCL3")
if (!is.null(goi)) {
  goi <- str_split(goi,",")[[1]]
  goi <- data[data$gene_name == goi,]
}

# Set hadjpval
hadjpval <- (-log10(max(
  data$P.Value[data$adj.P.Val < FDRq], 
  na.rm=TRUE)))

# Volcano plot
p <-
  ggplot(data = data, 
         aes(x = logFC,  # x-axis is logFC
             y = -log10(P.Value),  # y-axis will be -log10 of P.Value
             color = color)) +  # color is based on factored color column
  geom_point(alpha = 0.8, size = 2) +  # create scatterplot, alpha makes points transparent
  theme_bw() +  # set color theme
  theme(legend.position = "none") +  # no legend
  scale_color_manual(values = c("red", "blue","grey")) +  # set factor colors
  labs(
    title = paste0(tissue," LPS vs Saline\n", cutoff),
    x = expression(log[2](FC)), # x-axis title
    y = expression(-log[10] ~ "(" ~ italic("p") ~ "-value)") # y-axis title
  ) +
  theme(axis.title.x = element_text(size = 10),
        axis.text.x = element_text(size = 10)) +
  theme(axis.title.y = element_text(size = 10),
        axis.text.y = element_text(size = 10)) +
  geom_hline(yintercept = hadjpval,  colour = "black", linetype = "dashed") +
  geom_vline(xintercept = c(LFC,-LFC), colour = "black", linetype = "dashed") +
  geom_text_repel(data = up10,
                  aes(x = logFC, y= -log10(P.Value), label = gene_name), 
                  color = "maroon", 
                  fontface="italic",
                  max.overlaps = getOption("ggrepel.max.overlaps", default = 30)
                  ) +
  geom_text_repel(data = down10,
                  aes(x = logFC, y= -log10(P.Value), label = gene_name), 
                  color = "navyblue", 
                  fontface="italic",
                  max.overlaps = getOption("ggrepel.max.overlaps", default = 30)
                  ) +
  if (!is.null(goi)) {
      geom_text_repel(data = goi,
                  aes(x = logFC, y= -log10(P.Value), label = gene_name), 
                  color = "black", 
                  fontface="italic",
                  max.overlaps = getOption("ggrepel.max.overlaps", default = 30)
                  )
  }

p

# Display number of DEGs
table(num_DEGs)
```

## Save updated rds
```{r}
#saveRDS(data, "LPS_brain_gene_table.rds")
#saveRDS(data, "LPS_kidney_gene_table.rds")
```

# User interface
The user will have the option to choose tissue (brain or kidney) and the genes
to label (top 10 up and top 10 down or individual gene list entered)
```{r}
# User interface
ui <- fluidPage(

  # app title
  titlePanel("Bulk RNA-sequencing Volcano Plot"),
  
  # sidebar layout containing input and output definitions
  sidebarLayout(
    
    # sidebar panel for inputs
    sidebarPanel(
      # input
      selectInput(inputId = "tissue", label = "Tissue",
                  choices = c("Brain","Kidney"),
                  selected = "Brain"),
      selectInput(inputId = "cutoff", label = "FDRq and LFC Cutoffs", 
                   choices = c("FDRq_0.01_LFC_0.0",
                               "FDRq_0.01_LFC_0.5",
                               "FDRq_0.01_LFC_1.0",
                               "FDRq_0.05_LFC_0.0",
                               "FDRq_0.05_LFC_0.5",
                               "FDRq_0.05_LFC_1.0",
                               "FDRq_0.10_LFC_0.0",
                               "FDRq_0.10_LFC_0.5",
                               "FDRq_0.10_LFC_1.0"),
                  selected = "FDRq_0.05_LFC_0.0"),
    ),
    
    # main panel for displaying output
    mainPanel(
      #output
      plotOutput(outputId = "distPlot")
    )
  ) # end sidebarLayout
) # end fluidPage
```

# Server
```{r}
# Server logic required for histogram
server <- function(input, output) {
  
  output$distPlot <- renderPlot({
  
    # user variables
    tissue <- input$tissue
    cutoff <- input$cutoff
    
    # static variables
    genes_plotted <- "Top 10"
    
    # load data
    path <- paste0("/research/labs/neurology/fryer/m214960/LPS_Pigs/bulkRNA/rObjects/",
                   "LPS_",tolower(tissue),"_gene_table.rds")
    data <- readRDS(path)
      
    # color column is equal to cutoff column
    data$color <- factor(data[,cutoff])
    
    # get FDRq and LFC values
    FDRq <- as.numeric(str_match(cutoff,"FDRq_(.+)_LFC_(.+)")[,2])
    LFC <- as.numeric(str_match(cutoff,"FDRq_(.+)_LFC_(.+)")[,3])
    
    # Subset top 10 up and down regulated genes
    up <- data[data$color == 1,]
    up10 <- up[1:10,]
    down <- data[data$color == 2,]
    down10 <- down[1:10,]
    
    # Set hadjpval
    hadjpval <- (-log10(max(
      data$P.Value[data$adj.P.Val < FDRq], 
      na.rm=TRUE)))
    
    # Volcano plot
    p <-
      ggplot(data = data, 
             aes(x = logFC,  # x-axis is logFC
                 y = -log10(P.Value),  # y-axis will be -log10 of P.Value
                 color = color)) +  # color is based on factored color column
      geom_point(alpha = 0.8, size = 2) +  # create scatterplot, alpha makes points transparent
      theme_bw() +  # set color theme
      theme(legend.position = "none") +  # no legend
      scale_color_manual(values = c("red", "blue","grey")) +  # set factor colors
      labs(
        title = paste0(tissue," LPS vs Saline\n", cutoff),
        x = expression(log[2](FC)), # x-axis title
        y = expression(-log[10] ~ "(" ~ italic("p") ~ "-value)") # y-axis title
      ) +
      theme(axis.title.x = element_text(size = 10),
            axis.text.x = element_text(size = 10)) +
      theme(axis.title.y = element_text(size = 10),
            axis.text.y = element_text(size = 10)) +
      geom_hline(yintercept = hadjpval,  colour = "black", linetype = "dashed") +
      geom_vline(xintercept = c(LFC,-LFC), colour = "black", linetype = "dashed") +
      geom_text_repel(data = up10,
                      aes(x = logFC, y= -log10(P.Value), label = gene_name), 
                      color = "maroon", 
                      fontface="italic",
                      max.overlaps = getOption("ggrepel.max.overlaps", default = 30)
                      ) +
      geom_text_repel(data = down10,
                      aes(x = logFC, y= -log10(P.Value), label = gene_name), 
                      color = "navyblue", 
                      fontface="italic",
                      max.overlaps = getOption("ggrepel.max.overlaps", default = 30)
                      )
    p
  }) # end distPlot
}
```

# Run app
```{r}
# create shiny app
shinyApp(ui, server)
```




