---
title: "LPS Pigs Bulk RNAseq"
output: html_document
---

```{r load_packages,echo=FALSE,eval=TRUE,warning=FALSE,message=FALSE}
options(repos = BiocManager::repositories())
library(shiny)
library(ggplot2)
library(ggrepel)
library(rsconnect)
library(stringr)
library(BiocManager)
library(limma)
```

```{r,echo=FALSE,eval=TRUE,warning=FALSE,message=FALSE}
error.photo <- "
▂╱▔▔╲╱▔▔▔▔╲╱▔▔╲▂
╲┈▔╲┊╭╮┈┈╭╮┊╱▔┈╱
┊▔╲╱▏┈╱▔▔╲┈▕╲╱▔┊
┊┊┊┃┈┈▏┃┃▕┈┈┃┊┊┊
┊┊┊▏╲┈╲▂▂╱┈╱▕┊┊┊
"
# user choices (gtf used for featureCount)
gene_options <- readRDS("gene_options.rds")
```

```{r,echo=FALSE,eval=TRUE,warning=FALSE,message=FALSE}
plotVolcano <- function(tissue, FDRq, LFC, volcano_labels) {
  
  # set variable
  cutoff <- paste("FDRq",FDRq,"LFC",LFC,sep = "_")
  FDRq <- as.numeric(FDRq)
  LFC <- as.numeric(LFC)
  
  # load data
  data <- readRDS(paste0("LPS_",tolower(tissue),"_gene_table.rds"))
  gene_names <- data$gene_name
    
  # color column is equal to cutoff
  data$color <- factor(data[,cutoff])
  
  # Subset top 10 up and down regulated genes
  up <- data[data$color == 1,]
  up10 <- up[1:10,]
  down <- data[data$color == 2,]
  down10 <- down[1:10,]
  
  print(volcano_labels)
  # additional volcano labels
  if (!is.null(volcano_labels)) {
    # if gene is already labeled or filtered out of analysis
    if(any(volcano_labels %in% c(up10$gene_name,down10$gene_name))) {
      stop(safeError(
        paste0("One of the genes you selected is already labeled in the volcano.",
               error.photo)))
    } else if (!all(volcano_labels %in% gene_names)) {
      stop(safeError(
        paste0("One of the genes you selected was filtered out of the analysis",
               " due to low expression.",
               error.photo)))
    } else {
      additional_labels <- data[gene_names %in% volcano_labels,]
    }
  }
  
  # Set hadjpval
  hadjpval <- (-log10(max(
    data$P.Value[data$adj.P.Val < FDRq], 
    na.rm=TRUE)))
  
  # Volcano plot
  p <-
    ggplot(data = data, 
  		 aes(x = logFC,  # x-axis is LFC
  			 y = -log10(P.Value),  # y-axis will be -log10 of P.Value
  			 color = color)) +  # color is based on factored color column
    geom_point(alpha = 0.8, size = 2) +  # create scatterplot, alpha makes points transparent
    theme_bw() +  # set color theme
    theme(legend.position = "none") +  # no legend
    scale_color_manual(values = c("red", "blue","grey")) +  # set factor colors
    labs(
  	title = paste0(tissue," LPS vs Saline\nFDRq < ", FDRq, " LFC = ", LFC),
  	x = expression(log[2](FC)), # x-axis title
  	y = expression(-log[10] ~ "(" ~ italic("p") ~ "-value)") # y-axis title
    ) +
    theme(axis.title.x = element_text(size = 10),
  		axis.text.x = element_text(size = 10)) +
    theme(axis.title.y = element_text(size = 10),
  		axis.text.y = element_text(size = 10)) +
    geom_hline(yintercept = hadjpval,  colour = "black", linetype = "dashed") +
    geom_vline(xintercept = c(LFC,-LFC), colour = "black", linetype = "dashed") +
    geom_text_repel(data = up10,
  				  aes(x = logFC, y= -log10(P.Value), label = gene_name), 
  				  color = "maroon", 
  				  fontface="italic",
  				  max.overlaps = getOption("ggrepel.max.overlaps", default = 30)
  				  ) +
    geom_text_repel(data = down10,
  				  aes(x = logFC, y= -log10(P.Value), label = gene_name), 
  				  color = "navyblue", 
  				  fontface="italic",
  				  max.overlaps = getOption("ggrepel.max.overlaps", default = 30)
  				  ) +
    if (!is.null(volcano_labels)) {
  	  geom_text_repel(data = additional_labels,
  				  aes(x = logFC, y= -log10(P.Value), label = gene_name), 
  				  color = "black", 
  				  fontface="italic",
  				  max.overlaps = getOption("ggrepel.max.overlaps", default = 30)
  				  )
    }
  p
}
```

```{r,echo=FALSE,eval=TRUE,warning=FALSE,message=FALSE}
plotBoxplot <- function(tissue,gene_boxplot) {
    
  # load data
  data <- readRDS(paste0("LPS_",tolower(tissue),"_gene_voom.rds"))
  gene_names <- data$gene$gene_name
  counts <- data$E
  
  
  # throw error
  if(!gene_boxplot %in% gene_names) {
    stop(safeError(paste0(
      gene_boxplot,
      " was filtered out of the analysis due to low expression.\n",
      error.photo)))
  }

  # create df
  df <- as.data.frame(factor(data$targets$group))
  value <- subset(counts, rownames(counts) == gene_boxplot)
  value <- as.vector(value)
  colnames(df) <- "group"
  df$value <- value
  df <- df[c(2:5,7,8,1,6,9,10),]
  rownames(df) <- 1:nrow(df)
    
  # Visualize the distribution of genes detected per sample via boxplot
  sample_colors <- c("gray","purple")
  b <- ggplot(df, aes(x = group, y = value, fill = group)) +
    geom_boxplot(outlier.shape = NA) +
    geom_jitter() + 
    theme_classic() +
    theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust=1)) +
    theme(plot.title = element_text(hjust = 0.5, face="bold")) +
    ggtitle(gene_boxplot) +
    scale_color_manual(values = sample_colors) +
    scale_fill_manual(values = sample_colors) +
    ylab("logCPM value")
  b
}
```

```{r user_interface, echo=FALSE, eval=TRUE}
# User interface
ui <- fluidPage(

  # app title
  titlePanel("Gene level analysis with STAR"),
  
  # sidebar layout containing input and output definitions
  sidebarLayout(
    
    # sidebar panel for inputs
    sidebarPanel(
      
      # input
      selectInput(inputId = "tissue", label = "Select tissue",
                  choices = c("Brain","Kidney"),
                  selected = "Brain"),
      selectInput(inputId = "fdrq", label = "Select FDRq cutoff",
                  choices = c("0.01","0.05","0.10"),
                  selected = "0.05"),
      selectInput(inputId = "lfc", label = "Select LFC cutoff",
                  choices = c("0.0","0.5","1.0"),
                  selected = "0.0"),
      selectizeInput(inputId = "volcano_labels", 
                     label = "Additional volcano gene labels",
                     choices = NULL, multiple = TRUE),
      selectizeInput(inputId = "gene_boxplot",
                     label = "Gene of interest for boxplot",
                     choices = NULL),
      
      # submit button
      actionButton("submit", label = "Submit")
    ),
    
    # main panel for displaying output
    mainPanel(
      #output
      plotOutput(outputId = "volcano"),
      plotOutput(outputId = "boxplot")
    )
    
  ) # end sidebarLayout
) # end fluidPage
```

```{r server, echo=FALSE, eval=TRUE}
# Server logic required for histogram
server <- function(input, output, session) {
  
  # server-side selectize
  updateSelectizeInput(session, 'volcano_labels', 
                       choices = gene_options, server = TRUE,
                       options = list(maxOptions = 5),
                       selected = NULL)
  updateSelectizeInput(session, 'gene_boxplot', 
                       choices = gene_options, server = TRUE,
                       options = list(maxOptions = 5),
                       selected = "ATF3")
  
  # reactive expression
  text_reactive <- eventReactive(
    eventExpr = input$submit,
    valueExpr = c(input$tissue,input$fdrq,input$lfc,input$volcano_labels,
                  input$gene_boxplot))
  
  output$volcano <- renderPlot({
    # get new values
    text_reactive()
    # plot
    plotVolcano(tissue = input$tissue,
                FDRq = input$fdrq,
                LFC = input$lfc,
                volcano_labels = input$volcano_labels)
  })
  
  output$boxplot <- renderPlot({
    # get new values
    text_reactive()
    # plot boxplot
    plotBoxplot(input$tissue, input$gene_boxplot)
  })
}
```

```{r run_app, eval=TRUE, echo=FALSE}
# create shiny app
shinyApp(ui, server, options = list(height = 800) )
```

```{r rsconnect, echo=FALSE, eval=FALSE}
rsconnect::deployApp(appDir = '/Users/m214960/Documents/shiny/LPS_bulk',
                     appFiles = c("LPS_brain_gene_table.rds", 
                                  "LPS_kidney_gene_table.rds",
                                  "bulk_gene_volcano.Rmd",
                                  "LPS_kidney_gene_voom.rds",
                                  "LPS_brain_gene_voom.rds",
                                  "gene_options.rds"),
                     upload = TRUE,
                     appName = "LPS_Bulk_RNAseq",
                     account = "fryerlab")
```


